File: solutions/java/exercise3/OrderProcessor.java
package workshop.exercise3;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Map;
import java.util.Set;

// This exercise starts from a working-but-messy implementation; the tests already
// pass before any changes. The refactor below preserves that behavior while
// extracting named constants/helpers for each step the tests exercise.
public class OrderProcessor {

    private static final Set<String> SUPPORTED_STATES = Set.of("OH", "AZ", "TX");

    private static final Map<String, BigDecimal> TAX_RATES_BY_STATE = Map.of(
        "OH", new BigDecimal("0.0575"),
        "AZ", new BigDecimal("0.056"),
        "TX", new BigDecimal("0.0625")
    );

    public OrderResult processOrder(BigDecimal amount, String state, boolean isMember, int loyaltyYears) {
        // Exercise 3, Test: processOrder_InvalidAmount_ReturnsFailure
        // Orders with a non-positive (or missing) amount are rejected before anything else.
        if (amount == null || amount.compareTo(BigDecimal.ZERO) <= 0) {
            return failure("Invalid Amount");
        }

        // Exercise 3, Test: processOrder_UnsupportedState_ReturnsFailure
        // Only orders shipping to a supported state can be processed.
        if (!SUPPORTED_STATES.contains(state)) {
            return failure("Unsupported State");
        }

        // Exercise 3, Test: processOrder_ValidNonMember_CalculatesTax
        // Apply the state-specific tax rate to the order amount.
        BigDecimal tax = amount.multiply(TAX_RATES_BY_STATE.get(state));

        // Exercise 3, Test: processOrder_MemberDiscounts_AppliesCorrectTiers
        // Members earn a loyalty discount that scales with tenure.
        BigDecimal discount = isMember ? amount.multiply(memberDiscountRate(loyaltyYears)) : BigDecimal.ZERO;

        BigDecimal finalAmount = amount.add(tax).subtract(discount).max(BigDecimal.ZERO);

        return new OrderResult(true, finalAmount.setScale(2, RoundingMode.HALF_UP), "Processed");
    }

    private static BigDecimal memberDiscountRate(int loyaltyYears) {
        if (loyaltyYears >= 5) return new BigDecimal("0.15");
        if (loyaltyYears >= 2) return new BigDecimal("0.05");
        return new BigDecimal("0.02");
    }

    private static OrderResult failure(String message) {
        return new OrderResult(false, BigDecimal.ZERO, message);
    }
}

File: solutions/csharp/Exercise3/OrderProcessor.cs
using System;
using System.Collections.Generic;

namespace Workshop.Exercise3;

public class OrderResult
{
    public bool IsSuccess { get; set; }
    public decimal FinalAmount { get; set; }
    public string Message { get; set; } = string.Empty;
}

public class OrderProcessor
{
    private static readonly HashSet<string> SupportedStates = new() { "OH", "AZ", "TX" };

    private static readonly Dictionary<string, decimal> TaxRatesByState = new()
    {
        ["OH"] = 0.0575m,
        ["AZ"] = 0.056m,
        ["TX"] = 0.0625m,
    };

    // This exercise starts from a working-but-messy implementation; the tests already
    // pass before any changes. The refactor below preserves that behavior while
    // extracting named constants/helpers for each step the tests exercise.
    public OrderResult ProcessOrder(decimal amount, string state, bool isMember, int loyaltyYears)
    {
        // Exercise 3, Test: ProcessOrder_InvalidAmount_ReturnsFailure
        // Orders with a non-positive amount are rejected before anything else.
        if (amount <= 0)
        {
            return Failure("Invalid Amount");
        }

        // Exercise 3, Test: ProcessOrder_UnsupportedState_ReturnsFailure
        // Only orders shipping to a supported state can be processed.
        if (!SupportedStates.Contains(state))
        {
            return Failure("Unsupported State");
        }

        // Exercise 3, Test: ProcessOrder_ValidNonMember_CalculatesCorrectTax
        // Apply the state-specific tax rate to the order amount.
        decimal tax = amount * TaxRatesByState[state];

        // Exercise 3, Test: ProcessOrder_MemberDiscounts_AppliesCorrectTiers
        // Members earn a loyalty discount that scales with tenure.
        decimal discount = isMember ? amount * MemberDiscountRate(loyaltyYears) : 0m;

        decimal finalAmount = Math.Max(amount + tax - discount, 0m);

        return new OrderResult
        {
            IsSuccess = true,
            FinalAmount = Math.Round(finalAmount, 2),
            Message = "Processed",
        };
    }

    private static decimal MemberDiscountRate(int loyaltyYears)
    {
        if (loyaltyYears >= 5) return 0.15m;
        if (loyaltyYears >= 2) return 0.05m;
        return 0.02m;
    }

    private static OrderResult Failure(string message) => new()
    {
        IsSuccess = false,
        FinalAmount = 0m,
        Message = message,
    };
}
